rfd/0157-aws-eks-discover.md

```diff
@@ -0,0 +1,181 @@
+`teleport-kube-agent`.
+
+### Enrolling Selected Clusters
+
+For the enrollment of just the selected clusters, users will be further given two choices:
+- Use EKS Access entries to enroll a cluster with just the API.
+- As a backup option - manually run a Helm command that we generated for them.
+
+For completely automated cluster enrollment through the API, the Helm Go SDK will be used -
+allowing us to avoid the need for an additional step to run Helm itself elsewhere. It would require
+running a job inside the target EKS cluster with a special image that has Helm on it, and this job
+would install the Teleport kube agent. By using the Helm Go SDK, we can run the installation directly from the Teleport
+process and don't need to maintain a new special image and EKS installation job.
+
+### Automatic Discovery and Enrollment
+
+Users will also be able to set up automatic discovery and enrollment of EKS clusters in the Discover UI.
+We will reuse the existing EKS discovery capability, although a new mode will be added to it - discovery through
+the AWS integration. EKS discovery in that mode will rely on the AWS integration, selected in the Discovery UI,
+to find clusters and then install the Helm chart on them. The algorithm will be as follows:
+- List EKS clusters through the AWS integration.
+- Cross-check to see which EKS clusters are already registered in the system, using labels we inject on Helm chart installation.
+- Once a new EKS cluster is found, create EKS access entry for Teleport AWS Integration role and associate cluster admin policy.
+- Using the created access entry, get access to the EKS cluster and install the kube agent through the Helm chart.
+- Remove access entry from the EKS cluster.
+
+Users will be able to filter EKS clusters for discovery by using labels.
+
+A dynamic discovery configuration will be created as a result of users setting up automatic EKS discovery in the Discover UI (see [RFD 125](https://github.com/gravitational/teleport/blob/master/rfd/0125-dynamic-auto-discovery-config.md)).
+This configuration can be picked up by any discovery service. In the Teleport cloud, the Discovery service will be available by default,
+meaning users will not need to perform any additional actions.
+
+### EKS Access Entries
+
+AWS has recently introduced a new authentication method for EKS clusters - "access entries". 
+These access entries can be created through the AWS API and give permissions to an IAM user or role.
+Permissions can be granted in two ways: either directly through the access entry by specifying username/groups for the 
+IAM role, or by associating one of the predefined access policies to the access entry.
+
+Teleport will create an access entry for the role AWS integration is running with, and then will associate
+cluster admin policy to this access entry. It will allow AWS integration to authenticate with the EKS cluster and 
```

Comment:
We should say why we need the [`cluster-admin` arn:aws:eks::aws:cluster-access-policy/AmazonEKSClusterAdminPolicy](arn:aws:eks::aws:cluster-access-policy/AmazonEKSClusterAdminPolicy) policy - i.e. we need to create the NS and cluster role for the agent to get pods and list services...

We should also say that there is a limitation that requires the EKS auth to have the API auth enabled. For older clusters (created before the new API release, the default option is CONFIGMAP which means that this process won't work unless the cluster is converted to CONFIGMAP_API setting) - should we also include that into the scope? 

For new clusters, the default is already `CONFIGMAP_API` so they will work correctly

Should we have other methods as backup or will we enforce the requirement of this new value to be API or CONFIGMAP_API?

Reply:
Added this information to the RFD 👍  ( https://github.com/gravitational/teleport/pull/35282/commits/a3f8cf266a9b01c51f05d2979894a3d03bcb6e43 ). Also, mentioned that we will only support EKS cluster accessible to Teleport (otherwise user can use manual instructions we generate)

> Should we have other methods as backup or will we enforce the requirement of this new value to be API or CONFIGMAP_API?

I think it's fine to have this as a requirement, it's very easy and painless to migrate clusters to the new API (they have intermediate mode, when both configmap and API work) and AWS said that they are going to remove old way, so they are serious about moving people to the new access (it's better anyway).